Remove restriction for displaying Popup on iOS

cat0363 opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->
This PR makes it possible to use Popup when the display source of Popup is not PageHandler.

 ### Description of Change ###

The SetElement method in MauiPopup.macios.cs imposes a restriction that the Parent type must be PageHandler.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    public void SetElement(IPopup element)
    {
	// Remove start
        if (element.Parent?.Handler is not PageHandler)	
        {
            throw new InvalidOperationException($"The {nameof(element.Parent)} must be of type {typeof(PageHandler)}.");
        }
	// Remove end
        
        omit ...
    }

By removing this restriction, we fix it so that the Popup can be displayed even if the Popup's parent is not a PageHandler.

However, removing the above constraint causes the following problem.

1. By making the background color of the page where the Popup is displayed semi-transparent, 
the background color of the Popup is dimmed to achieve the overlay. However, if the page 
from which the Popup is displayed is displayed modally, if the background color of the page 
displayed modally is made translucent, the page from which the page displayed modally is 
displayed will also be visible.

Below is the verification video.

https://github.com/CommunityToolkit/Maui/assets/125236133/fb0104e8-e9f0-439d-a3df-052437c3cb11

2. Closing the displayed Popup not only closes the Popup, but also closes the page from 
which it was displayed. As a result, it will transition to the display source page of 
the page that was displayed modally.

Below is the verification video.

https://github.com/CommunityToolkit/Maui/assets/125236133/4f0fdb45-9229-413f-9b7a-b08601c9fe84

To solve the first problem, stop making the page from which the Popup is displayed 
semi-transparent to achieve the overlay. Add a new UIView to Subview to overlay on the 
page where the Popup is displayed.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    UIView overlayView;

    public MauiPopup(IMauiContext mauiContext)
    {
        this.mauiContext = mauiContext ?? throw new ArgumentNullException(nameof(mauiContext));
        overlayView = new UIView();
    }

    public override void ViewWillDisappear(bool animated)
    {
        overlayView.RemoveFromSuperview();
        base.ViewWillDisappear(animated);
    }

    public override void ViewWillTransitionToSize(CGSize toSize, IUIViewControllerTransitionCoordinator coordinator)
    {
        base.ViewWillTransitionToSize(toSize, coordinator);

        coordinator.AnimateAlongsideTransition((IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // Before screen rotate
            if (ViewController?.View is UIView view)
            {
                overlayView.Frame = new CGRect(view.Frame.X, view.Frame.Y, view.Frame.Width, view.Frame.Height);
            }
        }, (IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // After screen rotate
            if (VirtualView is not null)
            {
                PopupExtensions.SetSize(this, VirtualView);
                PopupExtensions.SetLayout(this, VirtualView);
            }
        });
    }

    void SetDimmingBackgroundEffect()
    {
        if (ViewController?.View is UIView view)
        {
            overlayView.Bounds = view.Bounds;
            overlayView.Layer.RemoveAllAnimations();
            overlayView.Frame = view.Frame;
            overlayView.BackgroundColor = UIColor.Black.ColorWithAlpha(0.4f);
            view.AddSubview(overlayView);
        }
    }

By detecting the rotation of the screen and specifying the size of the frame 
after rotation before rotation, the overlay view is not animated by the resize 
animation. I'm inverting the width and height.

    coordinator.AnimateAlongsideTransition((IUIViewControllerTransitionCoordinatorContext obj) =>
    {
        // Before screen rotate
        if (ViewController?.View is UIView view)
        {
            overlayView.Frame = new CGRect(view.Frame.X, view.Frame.Y, view.Frame.Width, view.Frame.Height);
        }
    }, 

The above solves the first problem.

To solve the second problem, stop calling the ViewController's DismissViewControllerAsync method.

[src\CommunityToolkit.Maui.Core\Handlers\Popup\PopupHandler.macios.cs]

    public static async void MapOnClosed(PopupHandler handler, IPopup view, object? result)
    {
	// Remove start
        var vc = handler.PlatformView.ViewController;
        if (vc is not null)
        {
            await vc.DismissViewControllerAsync(true);
        }
	// Remove end
    }

If the popup is displayed from a non-modally displayed page, the above will not be a problem, 
but if the popup is displayed from a modally displayed page, it will be closed, including 
the modally displayed page.

In order to properly close only the Popup, you need to change it as follows.

[src\CommunityToolkit.Maui.Core\Handlers\Popup\PopupHandler.macios.cs]

    public static async void MapOnClosed(PopupHandler handler, IPopup view, object? result)
    {
        var pc = handler.PlatformView.PresentationController;
        if (pc is not null)
        {
            if (pc.PresentedViewController is UIViewController pvc)
            {
                await pvc.DismissViewControllerAsync(true);
            }
        }
    }

The above solves the second problem.

 <!-- Describe your changes here. This only needs to be brief as the linked issues below will already cover the detailed changes. -->

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #1347

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->

I first verified it with the source code provided in Issue #1347.
 Below are the verification results.

https://github.com/CommunityToolkit/Maui/assets/125236133/0247ef90-9085-480b-b05b-e5676f67c50c

You can see that the page from which the Popup is displayed is translucent and the page behind it is no longer visible.
Also, when you close the Popup, you can see that the display source of the Popup is no longer closed.

Next, I tried rotating the screen while displaying the Popup.

https://github.com/CommunityToolkit/Maui/assets/125236133/e5a5b730-bd5e-49fc-852b-9415297b6947

You can see that the overlay resize animation does not occur and the overlay is displayed at its after rotated size.

Next, I displayed the Popup from a non-modal page and verified it.

https://github.com/CommunityToolkit/Maui/assets/125236133/d91dba46-156a-485a-804d-765e0365792a

You can see that the Popup appears and closes as intended.

brminnick commented:
@cat0363 Just FYI - the latest PR we merged today introduced a merge conflict in `MauiPopup.macios.cs`

brminnick commented:
@cat0363 Just FYI - a PR we merged today introduced a merge conflict here in this PR

cat0363 commented:
@brminnick , I resolved the conflict and merged.

cat0363 commented on src/CommunityToolkit.Maui.Core/Views/Popup/MauiPopup.macios.cs:
```diff
@@ -10,6 +10,7 @@ namespace CommunityToolkit.Maui.Core.Views;
 public class MauiPopup : UIViewController
 {
 	readonly IMauiContext mauiContext;
+	UIView overlayView;
```
@brminnick , Sorry, I overlooked WeakReference.
I would like to change the implementation to use WeakReference.

cat0363 replied:
@brminnick , After changing to use WeakReference, overlayView may be destroyed at an unintended 
timing when Popup is repeatedly show and close. Does this mean it has been disposed of by the 
garbage collector?

brminnick replied:
Off of the top of my head, I don't think that's too much of a concern because we disconnect the Popup's Handler and dispose of it when `IPopup.Close()` is called: 
https://github.com/CommunityToolkit/Maui/blob/0ab0fe22f4e701d1f463e243f85c2ddb07354a6a/src/CommunityToolkit.Maui.Core/Handlers/Popup/PopupHandler.macios.cs#L14-L25

I haven't done any testing, though. If you found a scenario where `overlayView` is disposed before Popup is disposed, let me know!

cat0363 replied:
@brminnick , Thank you for your reply.

I made the changes below.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    readonly WeakReference<UIView> overlayViewReference = new(new UIView());

    internal UIView OverlayView
    {
        get => overlayViewReference.TryGetTarget(out var overlayView) ? overlayView : throw new ObjectDisposedException(nameof(OverlayView));
        set => overlayViewReference.SetTarget(value);
    }

    public override void ViewWillDisappear(bool animated)
    {
        OverlayView.RemoveFromSuperview();
        base.ViewWillDisappear(animated);
    }

    public override void ViewWillTransitionToSize(CGSize toSize, IUIViewControllerTransitionCoordinator coordinator)
    {
        coordinator.AnimateAlongsideTransition((IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // Before screen rotate
            if (ViewController?.View is UIView view)
            {
                OverlayView.Frame = new CGRect(view.Frame.X, view.Frame.Y, view.Frame.Width, view.Frame.Height);
            }
        }, (IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // After screen rotate
            if (VirtualView is not null)
            {
                PopupExtensions.SetSize(this, VirtualView);
                PopupExtensions.SetLayout(this, VirtualView);
            }
        });
        base.ViewWillTransitionToSize(toSize, coordinator);
    }

    void SetDimmingBackgroundEffect()
    {
        if (ViewController?.View is UIView view)
        {
            OverlayView.Bounds = view.Bounds;
            OverlayView.Layer.RemoveAllAnimations();
            OverlayView.Frame = view.Frame;
            OverlayView.BackgroundColor = UIColor.Black.ColorWithAlpha(0.4f);
            view.AddSubview(OverlayView);
        }
    }

Before the MapOnClosed method of PopupHandler.macios.cs is called, the ViewWillDisapper 
method of MauiPopup.macios.cs is called, but the OverlayView has already been disposed.

The current implementation of the MapOnClosed method looks like this:

[src\CommunityToolkit.Maui.Core\Handlers\Popup\PopupHandler.macios.cs]

    public static async void MapOnClosed(PopupHandler handler, IPopup view, object? result)
    {
        var pc = handler.PlatformView.PresentationController;
        if (pc is not null)
        {
            if (pc.PresentedViewController is UIViewController pvc)
            {
                await pvc.DismissViewControllerAsync(true);
            }
        }

        view.HandlerCompleteTCS.TrySetResult();

        handler.DisconnectHandler(handler.PlatformView);
    }

The page that displays the Popup creates an instance of the Popup each time it is displayed.

In the display source of the Popup, an instance of the Popup is generated each time the Popup 
is displayed, but the OverlayView has already been destroyed when the Popup is closed by 
tapping the outside of the Popup. The OverlayView was destroyed before the OnMapClosed method was called.

brminnick replied:
Thanks! Do you think we should to reinstantiate `OverlayView` if it is null somewhere in `Popup.Show()` / `Popup.ShowAsync()` code path?

This is just an honest question, not a suggestion as I haven't reviewed the code. @TRybina132 has volunteered to lead the review this PR. I just chimed in since I noticed that my PR adding `MemoryAnalysis` broke your PR.

cat0363 replied:
> Thanks! Do you think we should to reinstantiate OverlayView if it is null somewhere in Popup.Show() / Popup.ShowAsync() code path?

I added an OverlayView to show a dim background behind the Popup.
If you create a new OverlayView just because the OverlayView was destroyed while the overlay 
was being displayed, the OverlayView that was already placed on the screen will remain on 
the screen without being removed. I think it's not a solution.

Below is the verification video.

https://github.com/CommunityToolkit/Maui/assets/125236133/f0335e0c-ba45-4943-b057-e5b52976213c

I understand that this is not a review. Thank you for your time.
If I find out anything, I'll post it here.

cat0363 replied:
When WeakReference is introduced, Target becomes null irregularly.
It's a movement that can't be explained without the garbage collector working, 
but I don't know if the garbage collector actually worked or not.
Is it because timing that is not referenced always occurs somewhere ...

cat0363 replied:
@brminnick , I would like to wait until a concrete implementation method is established.
Although unrelated to this case, the iOS Popup cannot use instances to call the ShowPopupAsync method. 
I will raise an issue on this matter tomorrow. When the CreateControl method is called a second time, the Popup's Content has already been set to the ContentPage's Content in the first call, so the second time there is a warning.
I think the ViewWillAppear method is better for calling the SetDimmingBackgroundEffect method.
I think that the processing of the SetView method also needs to be reviewed. In any case, it is not implemented to reuse Popup instances.

cat0363 replied:
I am generating an overlay View when the Popup is displayed, but I want to delete this View when the 
Popup is closed, so I temporarily hold it in a class variable. The view for this overlay should be 
kept as a WeakReference, but no one will reference it long-term during the Popup's lifetime.
Therefore, it is immediately discarded by the garbage collector.

In this PR, we only need to get the view for the overlay when the Popup is closed and remove it 
properly, so we propose the following alternatives.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    UIView GetOverlayView(UIView view)
    {
        var overlayViewTag = new IntPtr(38483);
        var overlayView = view.Subviews.AsEnumerable()
            .Where(
                x => x.Tag == overlayViewTag
            )
            .FirstOrDefault();
        if (overlayView == null)
        {
            overlayView = new UIView();
            overlayView.Tag = overlayViewTag;
        }

        return overlayView;
    }

    void SetDimmingBackgroundEffect()
    {
        if (ViewController?.View is UIView view)
        {
            var overlayView = GetOverlayView(view);
            overlayView.Bounds = view.Bounds;
            overlayView.Layer.RemoveAllAnimations();
            overlayView.Frame = view.Frame;
            overlayView.BackgroundColor = UIColor.Black.ColorWithAlpha(0.4f);
            view.AddSubview(overlayView);
        }
    }

    public override void ViewWillTransitionToSize(CGSize toSize, IUIViewControllerTransitionCoordinator coordinator)
    {
        coordinator.AnimateAlongsideTransition((IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // Before screen rotate
            if (ViewController?.View is UIView view)
            {
                var overlayView = GetOverlayView(view);
                overlayView.Frame = new CGRect(view.Frame.X, view.Frame.Y, view.Frame.Width, view.Frame.Height);
            }
        }, (IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // After screen rotate
            if (VirtualView is not null)
            {
                PopupExtensions.SetSize(this, VirtualView);
                PopupExtensions.SetLayout(this, VirtualView);
            }
        });
        base.ViewWillTransitionToSize(toSize, coordinator);
    }

    public override void ViewWillDisappear(bool animated)
    {
        if (ViewController?.View is UIView view)
        {
            var overlayView = GetOverlayView(view);
            overlayView.RemoveFromSuperview();
            overlayView.Dispose();
        }
        base.ViewWillDisappear(animated);
    }

With this implementation, you don't need to keep the overlay view in a class variable.
I am using Tag to search Subviews for overlay view. 
I remembered that Tag was used in the implementation below and thought about it.

[src\CommunityToolkit.Maui.Core\Platform\StatusBar\StatusBar.ios.cs]

It would have been nice if Tag could be set to a string instead of a number, 
but it's a pity that it only has numbers... The value set for Tag is the value of 
StatusBar's Tag plus 1, and has no deep meaning.

pictos commented:
@cat0363 Thanks for this PR and all the others... But I don't think this is the correct solution, forcing the popup to open when we have a `PageHandler` is this way by design, this control doesn't support being opened for any other control type (Label, Entry, Button, ContentView, Popup, etc), so, I would say, your fix should be other... Or there's nothing here to fix. Maybe the fix should be to proxy to the `Application.Current.MainPage` when we don't find the current page, and I believe there's a merged PR that forces that

cat0363 commented:
@pictos , Thank you for your valuable opinion.
I was waiting for that opinion. I have a question for you.
I don't understand the design concept, so I'm sorry if it's an irrelevant question.

1. Could you tell me the specific reasons for the following restrictions? And what problems does the lack of this restrictions create? I want to know that.

    public void SetElement(IPopup element)
    {
        if (element.Parent?.Handler is not PageHandler)
        {
            throw new InvalidOperationException($"The {nameof(element.Parent)} must be of type {typeof(PageHandler)}.");
        }

        VirtualView = element;
        ModalPresentationStyle = UIModalPresentationStyle.Popover;

        _ = View ?? throw new InvalidOperationException($"{nameof(View)} cannot be null.");
        _ = VirtualView ?? throw new InvalidOperationException($"{nameof(VirtualView)} cannot be null.");

        var rootViewController = WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
        ViewController ??= rootViewController;
        SetDimmingBackgroundEffect();
    }

2. Why don't Android and Windows have similar restrictions? I think that restrictions should be placed on Android and Windows so that there is no difference between what works and what doesn't work on each platform.

As an aside, the results of verification in PR before WeakReference is applied are described below.

[Case 1]

Open popup from popup. The first popup is displayed on the top left, the second popup is displayed in the center.

https://github.com/CommunityToolkit/Maui/assets/125236133/a0f90542-ba1a-420e-8a25-48fabc235818

[Case 2]

Open a Popup with the click event of MyButton, which inherits the Button class.

    public class MyButton : Button
    {
        public MyButton()
        {
            this.Clicked += (sender, args) =>
            {
                var popup3 = new PopupPage3();
                Shell.Current.ShowPopup(popup3);
            };
        }
    }

https://github.com/CommunityToolkit/Maui/assets/125236133/9fb453a8-7fde-42ca-8ea3-4d011c1e012d

The above case works equally well on Android and Windows.

As a result of discussion, if it is determined that there is no need to fix it, I will withdraw this PR.

pictos commented:
@cat0363 there are no bad questions at all, just ask as you need (:. I'll do my best do answer your questions, since it has been a long long time since I wrote those lines...

> Could you tell me the specific reasons for the following restrictions? And what problems does the lack of this restrictions create? I want to know that.

iOS doesn't have a `popup` concept, like Android (Dialog) and Windows (Flyout) have, so we mimic the Popup on iOS creating a `ViewController` and presenting it as a `UIModalPresentationStyle.Popover` in order to that work the `parent` of the popup should be anything that behaves like a `Page`, so because of that there's a check to make sure the `element.Parent` handler is a `PageHandler`.

> Why don't Android and Windows have similar restrictions? I think that restrictions should be placed on Android and Windows so that there is no difference between what works and what doesn't work on each platform.

Since Android and Windows have the concept of Popup, everything is handled by each platform. So, the limitation on iOS is to make the popup implementation there as close as possible to other platforms.


As I mentioned, has been a long time since I wrote these lines, maybe something changes that we can modify this code. Or even if Apple releases a popup control, we can use that instead.

cat0363 commented:
@pictos , Thank you for answering my question.

> Since Android and Windows have the concept of Popup, everything is handled by each platform. So, the limitation on iOS is to make the popup implementation there as close as possible to other platforms.

I understand why Windows and Android don't impose restrictions. The concept is to bring the incapable side closer to the capable side as much as possible, rather than matching the incapable side to the capable side, isn't it?

>iOS doesn't have a popup concept, like Android (Dialog) and Windows (Flyout) have, so we mimic the Popup on iOS creating a ViewController and presenting it as a UIModalPresentationStyle.Popover in order to that work the parent of the popup should be anything that behaves like a Page, so because of that there's a check to make sure the element.Parent handler is a PageHandler.

Is it correct to recognize that there are things that cannot be done if it is not PageHandler?
I would like to know what it is and the role of the Popup's parent.

In the issue that led to this PR, the type of `element.Parent?.Handler` was `PhoneFlyoutPageRenderer`.
It was not `PageHandler`. Does it mean that there is something that `PhoneFlyoutPageRenderer` 
cannot achieve? 

If the answer is yes, then you can add a condition, but the `PhoneFlyoutPageRenderer `cannot be 
referenced as it is defined at:

`Microsoft.Maui.Controls.Handlers.Compatibility.PhoneFlyoutPageRenderer`

So the implementation would look something like this:

    if (element.Parent?.Handler is not PageHandler &&
        element.Parent?.Handler?.GetType().Name != "PhoneFlyoutPageRenderer")
    {
        throw new InvalidOperationException($"The {nameof(element.Parent)} must be of type {typeof(PageHandler)}.");
    }

Wouldn't it be possible to add allowable parents as above?

brminnick commented:
@cat0363 FYI - after merging #1361, it has caused a merge conflict on this PR

cat0363 commented:
@brminnick , I resolved the conflict. But, I got a build error after merging, I will fix it on Monday. 

This PR does not give a clear answer to the question. I need to know a clear reason why element.Parent must be a PageHandler.
We need to write and verify an example of what happens when element.Parent is not a PageHandler. However, I haven't been able to describe the NG cases.

I think MacCatalyst at least needs to be considered separately from iOS since it is limited to PageHandler in PR #1409.
The build error seems to occur because this PR does not take into account the fixes in PR #1409.

cat0363 commented:
For MacCatalyst, element.Parent must be a PageHandler, so we reinstated the restriction.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    public void SetElement(IPopup element)
    {
        if (element.Parent?.Handler is not PageHandler)
        {
            throw new InvalidOperationException($"The {nameof(element.Parent)} must be of type {typeof(PageHandler)}.");
        }

        VirtualView = element;
        ModalPresentationStyle = UIModalPresentationStyle.Popover;

        _ = View ?? throw new InvalidOperationException($"{nameof(View)} cannot be null.");
        _ = VirtualView ?? throw new InvalidOperationException($"{nameof(VirtualView)} cannot be null.");

    #if MACCATALYST
        var pagehandler = VirtualView.Parent.Handler as PageHandler;
        var rootViewController = pagehandler?.ViewController ?? WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
    #else
        var rootViewController = WindowStateManager.Default.GetCurrentUIViewController() ?? throw new InvalidOperationException($"{nameof(PageHandler.ViewController)} cannot be null.");
    #endif

        ViewController ??= rootViewController;
        SetDimmingBackgroundEffect();
    }

In the case of MacCatalyst, if you set overlayView.Frame = view.Frame, the overlay does not completely cover the original view, so I fixed it as follows. For MacCatalyst, this is because there is an offset in the Y coordinate.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    void SetDimmingBackgroundEffect()
    {
        if (ViewController?.View is UIView view)
        {
            var overlayView = GetOverlayView(view);
            overlayView.Bounds = view.Bounds;
            overlayView.Layer.RemoveAllAnimations();
            overlayView.Frame = new CGRect(0, 0, view.Frame.Width, view.Frame.Height);
            overlayView.BackgroundColor = UIColor.Black.ColorWithAlpha(0.4f);
            view.AddSubview(overlayView);
        }
    }

    public override void ViewWillTransitionToSize(CGSize toSize, IUIViewControllerTransitionCoordinator coordinator)
    {
        coordinator.AnimateAlongsideTransition((IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // Before screen rotate
            if (ViewController?.View is UIView view)
            {
                var overlayView = GetOverlayView(view);
                overlayView.Frame = new CGRect(0, 0, view.Frame.Width, view.Frame.Height);
            }
        }, (IUIViewControllerTransitionCoordinatorContext obj) =>
        {
            // After screen rotate
            if (VirtualView is not null)
            {
                PopupExtensions.SetSize(this, VirtualView);
                PopupExtensions.SetLayout(this, VirtualView);
            }
        });
        base.ViewWillTransitionToSize(toSize, coordinator);
    }

Changed null judgment from ==null to is null.

[src\CommunityToolkit.Maui.Core\Views\Popup\MauiPopup.macios.cs]

    UIView GetOverlayView(UIView view)
    {
        var overlayViewTag = new IntPtr(38483);
        var overlayView = view.Subviews.AsEnumerable()
            .Where(
                x => x.Tag == overlayViewTag
            )
            .FirstOrDefault();
        if (overlayView is null)
        {
            overlayView = new UIView();
            overlayView.Tag = overlayViewTag;
        }

        return overlayView;
    }

brminnick reviewed (approved):
Great improvements! Thanks @cat0363!!
